core/ast_node.js

```diff
@@ -0,0 +1,657 @@
+  }
+  return newAstNode;
+};
+
+/**
+ * Navigate between stacks of blocks on the workspace.
+ * @param {?Boolean} forward True to go forward. False to go backwards.
+ * @return {?Blockly.BlockSvg} The first block of the next stack.
+ * @private
+ */
+Blockly.ASTNode.prototype.navigateBetweenStacks_ = function(forward) {
+  var curLocation = this.getLocation();
+  if (!(curLocation instanceof Blockly.Block)) {
+    curLocation = curLocation.sourceBlock_;
+  }
+  if (!curLocation) {
+    return null;
+  }
+  var curRoot = curLocation.getRootBlock();
+  var topBlocks = curRoot.workspace.getTopBlocks();
+  for (var i = 0; i < topBlocks.length; i++) {
+    var topBlock = topBlocks[i];
+    if (curRoot.id == topBlock.id) {
+      var offset = forward ? 1 : -1;
+      var resultIndex = i + offset;
+      if (resultIndex == -1) {
+        resultIndex = topBlocks.length - 1;
+      } else if (resultIndex == topBlocks.length) {
+        resultIndex = 0;
+      }
+      return topBlocks[resultIndex];
+    }
+  }
+  throw Error('Couldn\'t find ' + (forward ? 'next' : 'previous') +
+      ' stack?!?!?!');
+};
+
+/**
+ * Create an ast node from the top connection on the block.
+ * @param {Blockly.Block} block The block that we want to find the top
+ *     connection on.
+ * @return {?Blockly.ASTNode} The astnode containing the top connection.
+ * @private
+ */
+Blockly.ASTNode.prototype.findTopASTConnection_ = function(block) {
+  var previousConnection = block.previousConnection;
+  var outputConnection = block.outputConnection;
+  var newAstNode;
+  if (previousConnection) {
+    newAstNode = Blockly.ASTNode.createConnectionNode(previousConnection);
+  } else if (outputConnection) {
+    newAstNode = Blockly.ASTNode.createConnectionNode(outputConnection);
+  }
+  return newAstNode;
+};
+
+
+/**
+ * Given a location in a stack of blocks find the next out connection. If the
+ * location is nested the next out location should be the connected input.
+ * @param {!Blockly.Block} block The source block for the current location.
+ * @return {?Blockly.ASTNode} The ast node holding thenext out connection or
+ *     block.
+ * @private
+ */
+Blockly.ASTNode.prototype.findOutLocationForBlock_ = function(block) {
+  var newLocation, newAstNode;
+
+  newLocation = block.previousConnection.targetConnection;
+  newAstNode = Blockly.ASTNode.createConnectionNode(newLocation);
+  if (!newLocation) {
+    newLocation = block.previousConnection;
+    newAstNode = Blockly.ASTNode.createStackNode(block);
+  }
+  return newAstNode;
+};
+
+/**
+ * Find the element to the right of the current element in the ast.
+ * @return {?Blockly.ASTNode} An ast node that wraps the next field, connection,
+ *     block, or workspace. Return null if there is no value to the right.
+ */
+Blockly.ASTNode.prototype.next = function() {
+  var newAstNode = null;
+
+  switch (this.type_) {
+    case Blockly.ASTNode.types.WORKSPACE:
+      var newX = this.wsCoordinate_.x + Blockly.ASTNode.wsMove_;
```

Comment:
Add a TODO to limit this. The view is bounded to half a screen beyond the furthest blocks, but there's no easy way to translate that to Workspace coordinates right now (hence TODO).

Reply:
Done. Changes can be found in cursor_ast_update branch.